File: 2015/3/rust/src/main.rs
use std::collections::HashSet;
use std::fs;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Position {
    x: i32,
    y: i32,
}

impl Position {
    fn new() -> Self {
        Position { x: 0, y: 0 }
    }

    fn move_direction(&self, direction: Direction) -> Self {
        let (dx, dy) = direction.as_offset();
        Position {
            x: self.x + dx,
            y: self.y + dy,
        }
    }
}

#[derive(Debug, Clone, Copy)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

impl Direction {
    fn from_char(c: char) -> Option<Self> {
        match c {
            '^' => Some(Direction::Up),
            'v' => Some(Direction::Down),
            '<' => Some(Direction::Left),
            '>' => Some(Direction::Right),
            _ => None,
        }
    }

    fn as_offset(&self) -> (i32, i32) {
        match self {
            Direction::Up => (0, 1),
            Direction::Down => (0, -1),
            Direction::Left => (-1, 0),
            Direction::Right => (1, 0),
        }
    }
}

struct DeliverySystem {
    visited: HashSet<Position>,
}

impl DeliverySystem {
    fn new() -> Self {
        let mut visited = HashSet::new();
        visited.insert(Position::new());
        DeliverySystem { visited }
    }

    fn deliver_presents(&mut self, movements: &str) -> usize {
        let mut current_pos = Position::new();
        
        for c in movements.chars() {
            if let Some(direction) = Direction::from_char(c) {
                current_pos = current_pos.move_direction(direction);
                self.visited.insert(current_pos);
            }
        }
        
        self.visited.len()
    }

    fn deliver_presents_with_robo(&mut self, movements: &str) -> usize {
        let mut santa_pos = Position::new();
        let mut robo_pos = Position::new();
        
        for (i, c) in movements.chars().enumerate() {
            if let Some(direction) = Direction::from_char(c) {
                let current_pos = if i % 2 == 0 {
                    santa_pos = santa_pos.move_direction(direction);
                    santa_pos
                } else {
                    robo_pos = robo_pos.move_direction(direction);
                    robo_pos
                };
                self.visited.insert(current_pos);
            }
        }
        
        self.visited.len()
    }
}

fn read_input(file_path: &str) -> Result<String, std::io::Error> {
    fs::read_to_string(file_path)
}

fn main() -> Result<(), std::io::Error> {
    let movements = read_input("../input.txt")?;

    // Part 1: Santa alone
    let mut delivery_system = DeliverySystem::new();
    let houses_visited = delivery_system.deliver_presents(&movements);
    println!("Part 1: {}", houses_visited);

    // Part 2: Santa and Robo-Santa
    let mut delivery_system = DeliverySystem::new();
    let houses_visited = delivery_system.deliver_presents_with_robo(&movements);
    println!("Part 2: {}", houses_visited);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_santa() {
        let mut delivery_system = DeliverySystem::new();
        assert_eq!(delivery_system.deliver_presents(">"), 2);
        
        let mut delivery_system = DeliverySystem::new();
        assert_eq!(delivery_system.deliver_presents("^>v<"), 4);
    }

    #[test]
    fn test_santa_and_robo() {
        let mut delivery_system = DeliverySystem::new();
        assert_eq!(delivery_system.deliver_presents_with_robo("^v"), 3);
        
        let mut delivery_system = DeliverySystem::new();
        assert_eq!(delivery_system.deliver_presents_with_robo("^>v<"), 3);
    }

    #[test]
    fn test_position_movement() {
        let pos = Position::new();
        assert_eq!(pos.move_direction(Direction::Right), Position { x: 1, y: 0 });
        assert_eq!(pos.move_direction(Direction::Up), Position { x: 0, y: 1 });
    }
}

File: 2015/3/go/main.go
package main

import (
	"fmt"
	"os"
)

type Position struct {
	x int
	y int
}

type PositionKey struct {
	x, y int
}

type positionSet map[PositionKey]struct{}

func (s positionSet) add(pos Position) {
	key := PositionKey{x: pos.x, y: pos.y}
	s[key] = struct{}{}
}

func (s positionSet) has(pos Position) bool {
	key := PositionKey{x: pos.x, y: pos.y}
	_, ok := s[key]
	return ok
}

func calculate_unique_houses(movements string) (int, error) {
	current := Position{x: 0, y: 0}

	visited := make(positionSet)
	visited.add(current)

	directions := map[rune]Position{
		'>': {x: 1, y: 0},
		'<': {x: -1, y: 0},
		'^': {x: 0, y: 1},
		'v': {x: 0, y: -1},
	}
	for _, move := range movements {
		dir, ok := directions[move]
		if !ok {
			return 0, fmt.Errorf("invalid direction %c", move)
		}
		current.x += dir.x
		current.y += dir.y
		visited.add(current)
	}
	return len(visited), nil
}

func calculate_unique_houses_robo_santa(movements string) (int, error) {
	santaPos := Position{x: 0, y: 0}
	roboPos := Position{x: 0, y: 0}

	visited := make(positionSet)
	visited.add(santaPos)

	directions := map[rune]Position{
		'>': {x: 1, y: 0},
		'<': {x: -1, y: 0},
		'^': {x: 0, y: 1},
		'v': {x: 0, y: -1},
	}
	santaTurn := true
	for _, move := range movements {
		dir, ok := directions[move]
		if !ok {
			return 0, fmt.Errorf("invalid direction %c", move)
		}
		if santaTurn {
			santaPos.x += dir.x
			santaPos.y += dir.y
			visited.add(santaPos)
		} else {
			roboPos.x += dir.x
			roboPos.y += dir.y
			visited.add(roboPos)
		}
		santaTurn = !santaTurn
	}
	return len(visited), nil
}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file %v\n", err)
		os.Exit(1)
	}
	houses, err := calculate_unique_houses(string(data))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error solving Part 1 %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 1: %d\n", houses)

	robo, err := calculate_unique_houses_robo_santa(string(data))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error solving Part 2 %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 1: %d\n", robo)
}
